client/rpc/debug/src/lib.rs

```diff
@@ -57,49 +96,89 @@ where
 	C::Api: DebugRuntimeApi<B>,
 	C::Api: EthereumRuntimeRPCApi<B>,
 {
-	fn trace_transaction(
-		&self,
+	pub fn task(
+		client: Arc<C>,
+		backend: Arc<BE>,
+		frontier_backend: Arc<fc_db::Backend<B>>,
+		permit_pool: Arc<Semaphore>,
+	) -> (impl Future<Output = ()>, DebugRequester) {
+		let (tx, mut rx): (DebugRequester, _) =
+			sp_utils::mpsc::tracing_unbounded("debug-requester");
+
+		let fut = async move {
+			loop {
+				if let Some(((transaction_hash, params), response_tx)) = rx.next().await {
+					let client = client.clone();
+					let backend = backend.clone();
+					let frontier_backend = frontier_backend.clone();
+					let permit_pool = permit_pool.clone();
+					// Spawn the long running task in another thread.
+					let res = tokio::task::spawn_blocking(move || {
```

Comment:
Why not use 2 pools (the one that already exists in RPC, and one we use to handle debug request). The architecture would be:
(RPC pool - 4 threads) -> receive request, and add push it to a channel 
(Debug pool - 10 threads) -> read the channel and process the debug request

There is no need for locking (channel takes care of it)
We can configure the amount of threads independently of the RPC pool
RPC stays responsive even if all the Debug threads are taken
We can limit the queue of Debug request by setting a limit to the channel buffer

Reply:
If I understand correctly, what you are suggesting is to use a channel with bounded capacity instead of unbounded and have this capacity control the number of concurrent blocking tasks. Something like:

- Setup an dedicated bounded channel for communicating the rpc request and the blocking task handler. 

Then,

- Receive an rpc request.
- Send a message down that channel.
- Receive the message on the blocking task handler end.
- The channel receiver will either resolve the future or block here awaiting to resolve it if there is no available slots in the buffer.
- Once the message future is resolved, process the task **and finally** free the slot on the channel buffer.

The problem is afaik (@nanocryk ?) the [receiver.recv](https://docs.rs/tokio/0.2.23/tokio/sync/mpsc/struct.Receiver.html#method.recv) only blocks (and reserves a capacity slot) until receving a message down the channel, after that the slot is freed and accepts new messages.

The solution for that would be the one proposed in this PR: instead we use an unbounded channel for rpc request <> task communication and a `Semaphore`, with its own capacity and ability to acquire/reserve a slot for **a scope** (the one that spawns the blocking task) instead. Additionally the Semaphore's permit pool can be shared between modules and a bounded channel has a single consumer/receiver (unless we use a broadcast channel).

Reply:
It is not really what I suggested (which is more a generic pattern and less tokio-friendly from what @nanocryk told me)
, but I'm fine using the semaphore technique if done properly.